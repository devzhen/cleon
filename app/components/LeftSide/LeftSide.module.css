.leftSide {
  position: relative;
  display: grid;
  grid-template-rows: 1fr 1fr 1fr;
  grid-template-columns: 1fr 1fr 1fr;
  min-width: 638px;
  max-width: 638px;
  min-height: 863px;
  height: calc(100vh - 72px);
  border: 3px solid var(--color-active-board);
  background-color: var(--color-board-background);
  z-index: 2;
}

.leftSide [data-role='drag-area'] {
  position: absolute;
}

.leftBannerWrapper {
  padding-top: 80px;
}

.logoWrapper {
  display: flex;
  justify-content: center;
  align-items: flex-start;
}

.logoWrapper div {
  display: flex;
  align-items: center;
  user-select: none;
  margin-top: 8px;
}

.logoWrapper div span {
  margin-left: 8px;
  color: var(--color-active-board);
  text-transform: uppercase;
  font-size: 28px;
}

.buttonsWrapper {
  display: flex;
  justify-content: flex-end;
  align-items: flex-start;
}

.bottomBannerWrapper {
  display: flex;
  justify-content: flex-end;
  align-items: flex-end;
}

.floatingWrapper {
  width: fit-content;
  height: fit-content;
}

.banner {
  width: 185px;
  height: 132px;
  margin-left: 10px;
  user-select: none;
}

.banner img {
  cursor: pointer;
}

.leftSide .banner:nth-child(2),
.leftSide .banner:nth-child(3) {
  margin-top: 10px;
}

.bottomBanner {
  margin: 0 10px 10px 0;
  width: 129px;
  height: 188px;
}
