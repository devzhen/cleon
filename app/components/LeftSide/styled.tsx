import styled from "styled-components";

export const Container = styled.div`
  position: relative;
  display: grid;
  grid-template-rows: 1fr 1fr 1fr;
  grid-template-columns: 1fr 1fr 1fr;
  min-width: 638px;
  max-width: 638px;
  min-height: 863px;
  height: calc(100vh - 72px);
  border: 3px solid ${(props) => props.theme.colors.activeBoard};
  background-color: ${(props) => props.theme.colors.boardBackground};
  z-index: 2;

  .drag-area {
    position: absolute;
  }
`;

export const Cell = styled.div<{
  $alignItems?: string;
  $justifyContent?: string;
}>`
  display: flex;

  ${(props) => props.$alignItems && `align-items: ${props.$alignItems}`};
  ${(props) =>
    props.$justifyContent && `justify-content: ${props.$justifyContent}`};
`;

export const FloatingWrapper = styled.div<{
  $paddingTop?: number;
}>`
  width: fit-content;
  height: fit-content;

  ${(props) => props.$paddingTop && `padding-top: ${props.$paddingTop}px`};
`;

export const Banner = styled.div<{
  $marginTop?: number;
  $marginLeft?: number;
  $marginRight?: number;
  $marginBottom?: number;
  $width?: number;
  $height?: number;
}>`
  width: ${(props) => props.$width || 185}px;
  height: ${(props) => props.$height || 132}px;
  user-select: none;

  ${(props) => props.$marginTop && `margin-top: ${props.$marginTop}px`};
  ${(props) => props.$marginLeft && `margin-left: ${props.$marginLeft}px`};
  ${(props) => props.$marginRight && `margin-right: ${props.$marginRight}px`};
  ${(props) =>
    props.$marginBottom && `margin-bottom: ${props.$marginBottom}px`};

  img {
    cursor: pointer;
  }
`;

export const Button = styled.button<{
  $marginTop?: number;
}>`
  display: flex;
  align-items: center;
  justify-content: center;
  width: 57px;
  height: 31px;
  background-color: ${(props) => props.theme.colors.mainBackground};
  cursor: pointer;
  color: ${(props) => props.theme.colors.white};
  user-select: none;
  font-size: 30px;
  font-weight: 400;
  line-height: 20px;
  text-transform: uppercase;
  border: none;
  outline: none;

  &:active {
    box-shadow: 2px 2px 5px ${(props) => props.theme.colors.mainBackground};
  }

  ${(props) => props.$marginTop && `margin-top: ${props.$marginTop}px`};
`;

export const LogoWrapper = styled.div`
  display: flex;
  align-items: center;
  padding-top: 10px;
  user-select: none;

  span {
    margin-left: 8px;
    color: ${(props) => props.theme.colors.activeBoard};
    text-transform: uppercase;
    font-size: 28px;
  }
`;
