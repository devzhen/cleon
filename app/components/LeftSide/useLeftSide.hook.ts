import { MutableRefObject, useRef } from "react";

import getCoords from "@/app/utils/getCoords";
import isPointInPath from "@/app/utils/isPointInPath";

type UseLeftSideProps = {
  currentDragElementRef: MutableRefObject<HTMLDivElement | undefined>
}

const useLeftSide = (props: UseLeftSideProps) => {
  const { currentDragElementRef } = props;

  // Current draggable sticker
  const currentDragElementParentRef = useRef<HTMLDivElement>();

  // Drag containers
  const dragContainerRef = useRef<HTMLDivElement>(null);
  const dragArea1Ref = useRef<HTMLDivElement>(null);
  const dragArea2Ref = useRef<HTMLDivElement>(null);
  const dragArea3Ref = useRef<HTMLDivElement>(null);

  // Banners, logo, buttons
  const leftBannerWrapperRef = useRef<HTMLDivElement>(null);
  const logoWrapperRef = useRef<HTMLDivElement>(null);
  const buttonsWrapperRef = useRef<HTMLDivElement>(null);
  const lastBannerWrapperRef = useRef<HTMLDivElement>(null);

  /**
   * Init draggable areas
   */
  const initDragAreas = () => {
    if (
      !dragContainerRef.current ||
      !dragArea1Ref.current ||
      !dragArea2Ref.current ||
      !dragArea3Ref.current ||
      !leftBannerWrapperRef.current ||
      !logoWrapperRef.current ||
      !buttonsWrapperRef.current ||
      !lastBannerWrapperRef.current
    ) {
      return;
    }

    const dragContainerRect = dragContainerRef.current.getBoundingClientRect();

    const leftBannerWrapperRect =
      leftBannerWrapperRef.current.getBoundingClientRect();

    const logoWrapperRect = logoWrapperRef.current.getBoundingClientRect();

    const buttonsWrapperRect =
      buttonsWrapperRef.current.getBoundingClientRect();

    const lastBannerWrapperRect =
      lastBannerWrapperRef.current.getBoundingClientRect();

    const BORDER_WIDTH = 3;

    const dragArea1Rect = {
      left: leftBannerWrapperRect.width,
      top: logoWrapperRect.height,
      width:
        dragContainerRect.width -
        leftBannerWrapperRect.width -
        buttonsWrapperRect.width -
        BORDER_WIDTH * 2,
      height:
        dragContainerRect.height -
        (logoWrapperRect.height +
          lastBannerWrapperRect.height +
          BORDER_WIDTH * 2),
    };

    dragArea1Ref.current.style.top = `${dragArea1Rect.top}px`;
    dragArea1Ref.current.style.left = `${dragArea1Rect.left}px`;
    dragArea1Ref.current.style.width = `${dragArea1Rect.width}px`;
    dragArea1Ref.current.style.height = `${dragArea1Rect.height}px`;

    const dragArea2Rect = {
      left: leftBannerWrapperRect.width,
      top: buttonsWrapperRect.height,
      width:
        dragContainerRect.width -
        leftBannerWrapperRect.width -
        BORDER_WIDTH * 2,
      height:
        dragContainerRect.height -
        (buttonsWrapperRect.height +
          lastBannerWrapperRect.height +
          BORDER_WIDTH * 2),
    };

    dragArea2Ref.current.style.top = `${dragArea2Rect.top}px`;
    dragArea2Ref.current.style.left = `${dragArea2Rect.left}px`;
    dragArea2Ref.current.style.width = `${dragArea2Rect.width}px`;
    dragArea2Ref.current.style.height = `${dragArea2Rect.height}px`;

    const dragArea3Rect = {
      left: 0,
      top: leftBannerWrapperRect.height,
      width:
        dragContainerRect.width -
        lastBannerWrapperRect.width -
        BORDER_WIDTH * 2,
      height:
        dragContainerRect.height -
        (leftBannerWrapperRect.height + BORDER_WIDTH * 2),
    };

    dragArea3Ref.current.style.top = `${dragArea3Rect.top}px`;
    dragArea3Ref.current.style.left = `${dragArea3Rect.left}px`;
    dragArea3Ref.current.style.width = `${dragArea3Rect.width}px`;
    dragArea3Ref.current.style.height = `${dragArea3Rect.height}px`;

    currentDragElementParentRef.current = dragArea2Ref.current;
  };

  /**
   * Detect a sticker's parent container
   */
  const detectParentContainer = () => {
    if (
      !dragArea1Ref.current ||
      !dragArea2Ref.current ||
      !dragArea3Ref.current ||
      !leftBannerWrapperRef.current ||
      !logoWrapperRef.current ||
      !buttonsWrapperRef.current ||
      !lastBannerWrapperRef.current
    ) {
      return;
    }

    const stickerCoords = getCoords(
      currentDragElementRef.current as HTMLDivElement
    );

    const dragArea1Coords = getCoords(dragArea1Ref.current);
    const dragArea2Coords = getCoords(dragArea2Ref.current);
    const dragArea3Coords = getCoords(dragArea3Ref.current);

    // The drag area 1
    if (currentDragElementParentRef.current === dragArea1Ref.current) {
      // Attach to the drag area 2
      if (
        stickerCoords.top > dragArea2Coords.top &&
        !isPointInPath(
          leftBannerWrapperRef.current,
          stickerCoords.left,
          stickerCoords.top
        ) &&
        !isPointInPath(
          buttonsWrapperRef.current,
          stickerCoords.right,
          stickerCoords.top
        )
      ) {
        currentDragElementParentRef.current = dragArea2Ref.current;
      }
    }

    // The drag area 2
    if (currentDragElementParentRef.current === dragArea2Ref.current) {
      // Attach to the drag area 3
      if (
        stickerCoords.top > dragArea3Coords.top &&
        stickerCoords.left > dragArea3Coords.left &&
        stickerCoords.right < dragArea3Coords.right &&
        stickerCoords.bottom < dragArea3Coords.bottom
      ) {
        currentDragElementParentRef.current = dragArea3Ref.current;
      }

      // Attach to the drag area 1
      if (
        stickerCoords.top < dragArea2Coords.top &&
        stickerCoords.right <= dragArea1Coords.right
      ) {
        currentDragElementParentRef.current = dragArea1Ref.current;
      }
    }

    // The drag area 3
    if (currentDragElementParentRef.current === dragArea3Ref.current) {
      // If move to top, attach to the drag area 2
      if (
        stickerCoords.top < dragArea3Coords.top &&
        stickerCoords.left > dragArea3Coords.left &&
        stickerCoords.right < dragArea3Coords.right &&
        stickerCoords.bottom < dragArea3Coords.bottom &&
        !isPointInPath(
          leftBannerWrapperRef.current,
          stickerCoords.left,
          stickerCoords.top
        )
      ) {
        currentDragElementParentRef.current = dragArea2Ref.current;
      }

      // If move to right, attach to the drag area 2
      if (
        stickerCoords.top > dragArea3Coords.top &&
        stickerCoords.left > dragArea3Coords.left &&
        stickerCoords.right > dragArea3Coords.right &&
        stickerCoords.bottom < dragArea3Coords.bottom &&
        !isPointInPath(
          lastBannerWrapperRef.current,
          stickerCoords.right,
          stickerCoords.bottom
        )
      ) {
        currentDragElementParentRef.current = dragArea2Ref.current;
      }
    }
  };

  /**
   * Correct a sticker's position relative to a parent
   */
  const correctPositionRelativeToParent = () => {
    if (
      !currentDragElementRef.current ||
      !currentDragElementParentRef.current
    ) {
      return;
    }

    const stickerCoords = getCoords(
      currentDragElementRef.current as HTMLDivElement
    );

    const parentCoords = getCoords(
      currentDragElementParentRef.current as HTMLDivElement
    );

    if (stickerCoords.left < parentCoords.left) {
      currentDragElementRef.current.style.left = `${currentDragElementParentRef.current.offsetLeft}px`;
    }

    if (stickerCoords.top < parentCoords.top) {
      currentDragElementRef.current.style.top = `${currentDragElementParentRef.current.offsetTop}px`;
    }

    if (stickerCoords.left + stickerCoords.width > parentCoords.right) {
      currentDragElementRef.current.style.left = `${
        currentDragElementParentRef.current.offsetLeft +
        parentCoords.width -
        stickerCoords.width
      }px`;
    }

    if (stickerCoords.top + stickerCoords.width > parentCoords.bottom) {
      currentDragElementRef.current.style.top = `${
        currentDragElementParentRef.current?.offsetTop +
        parentCoords.height -
        stickerCoords.height
      }px`;
    }
  };

  /**
   * Correct a sticker's position relative to a drag container
   */
  const correctPositionRelativeToDragContainer = () => {
    if (!currentDragElementRef.current || !dragContainerRef.current) {
      return;
    }

    const stickerCoords = getCoords(
      currentDragElementRef.current as HTMLDivElement
    );

    const dragContainerCoords = getCoords(
      dragContainerRef.current as HTMLDivElement
    );

    if (stickerCoords.left < dragContainerCoords.left) {
      currentDragElementRef.current.style.left = `${0}px`;
    }

    if (stickerCoords.top < dragContainerCoords.top) {
      currentDragElementRef.current.style.top = `${0}px`;
    }

    if (stickerCoords.left + stickerCoords.width > dragContainerCoords.right) {
      currentDragElementRef.current.style.left = `${
        dragContainerCoords.width - stickerCoords.width
      }px`;
    }

    if (stickerCoords.top + stickerCoords.width > dragContainerCoords.bottom) {
      currentDragElementRef.current.style.top = `${
        dragContainerCoords.height - stickerCoords.height
      }px`;
    }
  };

  return {
    currentDragElementRef,
    currentDragElementParentRef,
    dragContainerRef,
    dragArea1Ref,
    dragArea2Ref,
    dragArea3Ref,
    leftBannerWrapperRef,
    logoWrapperRef,
    buttonsWrapperRef,
    lastBannerWrapperRef,
    initDragAreas,
    detectParentContainer,
    correctPositionRelativeToParent,
    correctPositionRelativeToDragContainer,
  };
}

export default useLeftSide;