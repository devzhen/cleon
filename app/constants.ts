export const DEFAULT_COORDS = {
  top: 100,
  left: 220,
} as const;

export const BOARD_TYPE = {
  left: 'left',
  right: 'right',
} as const;