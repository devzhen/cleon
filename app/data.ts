import { v4 as uuid } from "uuid";

export const stickers = [
  {
    id: uuid(),
    top: 110,
    left: 230,
    createdAt: new Date().toISOString(),
    text: ''
  }
];
