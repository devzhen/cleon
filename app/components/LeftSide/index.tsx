import Image from "next/image";
import { useRef } from "react";

import { BOARD_TYPE } from "@/app/constants";
import useDraggableAreas from "@/app/hooks/useDraggableAreas";
import useSticker from "@/app/hooks/useSticker";
import getCoords from "@/app/utils/getCoords";
import isPointInPath from "@/app/utils/isPointInPath";

import ButtonsAddDelete from "../ButtonsAddDelete";
import ModalAddSticker from "../ModalAddSticker";
import StickerComponent from "../Sticker";

import {
  Container,
  Cell,
  FloatingWrapper,
  Banner,
  LogoWrapper,
} from "./styled";

export default function LeftSide() {
  // Current draggable sticker
  const currentDragElementRef = useRef<HTMLDivElement>();

  // Current draggable sticker's parent
  const currentDragElementParentRef = useRef<HTMLDivElement>();

  // Drag containers
  const dragContainerRef = useRef<HTMLDivElement>(null);
  const dragArea1Ref = useRef<HTMLDivElement>(null);
  const dragArea2Ref = useRef<HTMLDivElement>(null);
  const dragArea3Ref = useRef<HTMLDivElement>(null);

  // Banners, logo, buttons
  const leftBannerWrapperRef = useRef<HTMLDivElement>(null);
  const logoWrapperRef = useRef<HTMLDivElement>(null);
  const buttonsWrapperRef = useRef<HTMLDivElement>(null);
  const lastBannerWrapperRef = useRef<HTMLDivElement>(null);

  /**
   * Init draggable areas
   */
  const initDragAreas = () => {
    if (
      !dragContainerRef.current ||
      !dragArea1Ref.current ||
      !dragArea2Ref.current ||
      !dragArea3Ref.current ||
      !leftBannerWrapperRef.current ||
      !logoWrapperRef.current ||
      !buttonsWrapperRef.current ||
      !lastBannerWrapperRef.current
    ) {
      return;
    }

    const dragContainerRect = dragContainerRef.current.getBoundingClientRect();

    const leftBannerWrapperRect =
      leftBannerWrapperRef.current.getBoundingClientRect();

    const logoWrapperRect = logoWrapperRef.current.getBoundingClientRect();

    const buttonsWrapperRect =
      buttonsWrapperRef.current.getBoundingClientRect();

    const lastBannerWrapperRect =
      lastBannerWrapperRef.current.getBoundingClientRect();

    const BORDER_WIDTH = 3;

    const dragArea1Rect = {
      left: leftBannerWrapperRect.width,
      top: logoWrapperRect.height,
      width:
        dragContainerRect.width -
        leftBannerWrapperRect.width -
        buttonsWrapperRect.width -
        BORDER_WIDTH * 2,
      height:
        dragContainerRect.height -
        (logoWrapperRect.height +
          lastBannerWrapperRect.height +
          BORDER_WIDTH * 2),
    };

    dragArea1Ref.current.style.top = `${dragArea1Rect.top}px`;
    dragArea1Ref.current.style.left = `${dragArea1Rect.left}px`;
    dragArea1Ref.current.style.width = `${dragArea1Rect.width}px`;
    dragArea1Ref.current.style.height = `${dragArea1Rect.height}px`;

    const dragArea2Rect = {
      left: leftBannerWrapperRect.width,
      top: buttonsWrapperRect.height,
      width:
        dragContainerRect.width -
        leftBannerWrapperRect.width -
        BORDER_WIDTH * 2,
      height:
        dragContainerRect.height -
        (buttonsWrapperRect.height +
          lastBannerWrapperRect.height +
          BORDER_WIDTH * 2),
    };

    dragArea2Ref.current.style.top = `${dragArea2Rect.top}px`;
    dragArea2Ref.current.style.left = `${dragArea2Rect.left}px`;
    dragArea2Ref.current.style.width = `${dragArea2Rect.width}px`;
    dragArea2Ref.current.style.height = `${dragArea2Rect.height}px`;

    const dragArea3Rect = {
      left: 0,
      top: leftBannerWrapperRect.height,
      width:
        dragContainerRect.width -
        lastBannerWrapperRect.width -
        BORDER_WIDTH * 2,
      height:
        dragContainerRect.height -
        (leftBannerWrapperRect.height + BORDER_WIDTH * 2),
    };

    dragArea3Ref.current.style.top = `${dragArea3Rect.top}px`;
    dragArea3Ref.current.style.left = `${dragArea3Rect.left}px`;
    dragArea3Ref.current.style.width = `${dragArea3Rect.width}px`;
    dragArea3Ref.current.style.height = `${dragArea3Rect.height}px`;

    currentDragElementParentRef.current = dragArea2Ref.current;
  };

  /**
   * Detect a sticker's parent container
   */
  const detectParentContainer = () => {
    if (
      !dragArea1Ref.current ||
      !dragArea2Ref.current ||
      !dragArea3Ref.current ||
      !leftBannerWrapperRef.current ||
      !logoWrapperRef.current ||
      !buttonsWrapperRef.current ||
      !lastBannerWrapperRef.current
    ) {
      return;
    }

    const stickerCoords = getCoords(
      currentDragElementRef.current as HTMLDivElement
    );

    const dragArea1Coords = getCoords(dragArea1Ref.current);
    const dragArea2Coords = getCoords(dragArea2Ref.current);
    const dragArea3Coords = getCoords(dragArea3Ref.current);

    // The drag area 1
    if (currentDragElementParentRef.current === dragArea1Ref.current) {
      // Attach to the drag area 2
      if (
        stickerCoords.top > dragArea2Coords.top &&
        !isPointInPath(
          leftBannerWrapperRef.current,
          stickerCoords.left,
          stickerCoords.top
        ) &&
        !isPointInPath(
          buttonsWrapperRef.current,
          stickerCoords.right,
          stickerCoords.top
        )
      ) {
        currentDragElementParentRef.current = dragArea2Ref.current;
      }
    }

    // The drag area 2
    if (currentDragElementParentRef.current === dragArea2Ref.current) {
      // Attach to the drag area 3
      if (
        stickerCoords.top > dragArea3Coords.top &&
        stickerCoords.left > dragArea3Coords.left &&
        stickerCoords.right < dragArea3Coords.right &&
        stickerCoords.bottom < dragArea3Coords.bottom
      ) {
        currentDragElementParentRef.current = dragArea3Ref.current;
      }

      // Attach to the drag area 1
      if (
        stickerCoords.top < dragArea2Coords.top &&
        stickerCoords.right <= dragArea1Coords.right
      ) {
        currentDragElementParentRef.current = dragArea1Ref.current;
      }
    }

    // The drag area 3
    if (currentDragElementParentRef.current === dragArea3Ref.current) {
      // If move to top, attach to the drag area 2
      if (
        stickerCoords.top < dragArea3Coords.top &&
        stickerCoords.left > dragArea3Coords.left &&
        stickerCoords.right < dragArea3Coords.right &&
        stickerCoords.bottom < dragArea3Coords.bottom &&
        !isPointInPath(
          leftBannerWrapperRef.current,
          stickerCoords.left,
          stickerCoords.top
        )
      ) {
        currentDragElementParentRef.current = dragArea2Ref.current;
      }

      // If move to right, attach to the drag area 2
      if (
        stickerCoords.top > dragArea3Coords.top &&
        stickerCoords.left > dragArea3Coords.left &&
        stickerCoords.right > dragArea3Coords.right &&
        stickerCoords.bottom < dragArea3Coords.bottom &&
        !isPointInPath(
          lastBannerWrapperRef.current,
          stickerCoords.right,
          stickerCoords.bottom
        )
      ) {
        currentDragElementParentRef.current = dragArea2Ref.current;
      }

      // If move to right, attach to the drag area 2
      if (
        stickerCoords.top <= dragArea3Coords.top &&
        stickerCoords.left > dragArea3Coords.left &&
        stickerCoords.right >= dragArea3Coords.right &&
        stickerCoords.bottom < dragArea3Coords.bottom &&
        isPointInPath(
          lastBannerWrapperRef.current,
          stickerCoords.right,
          stickerCoords.bottom
        )
      ) {
        currentDragElementParentRef.current = dragArea2Ref.current;
      }
    }
  };

  /**
   * Correct a sticker's position relative to a parent
   */
  const correctPositionRelativeToParent = () => {
    if (
      !currentDragElementRef.current ||
      !currentDragElementParentRef.current
    ) {
      return;
    }

    const stickerCoords = getCoords(
      currentDragElementRef.current as HTMLDivElement
    );

    const parentCoords = getCoords(
      currentDragElementParentRef.current as HTMLDivElement
    );

    if (stickerCoords.left < parentCoords.left) {
      currentDragElementRef.current.style.left = `${currentDragElementParentRef.current.offsetLeft}px`;
    }

    if (stickerCoords.top < parentCoords.top) {
      currentDragElementRef.current.style.top = `${currentDragElementParentRef.current.offsetTop}px`;
    }

    if (stickerCoords.left + stickerCoords.width > parentCoords.right) {
      currentDragElementRef.current.style.left = `${
        currentDragElementParentRef.current.offsetLeft +
        parentCoords.width -
        stickerCoords.width
      }px`;
    }

    if (stickerCoords.top + stickerCoords.width > parentCoords.bottom) {
      currentDragElementRef.current.style.top = `${
        currentDragElementParentRef.current?.offsetTop +
        parentCoords.height -
        stickerCoords.height
      }px`;
    }
  };

  /**
   * Correct a sticker's position relative to a drag container
   */
  const correctPositionRelativeToDragContainer = () => {
    if (!currentDragElementRef.current || !dragContainerRef.current) {
      return;
    }

    const stickerCoords = getCoords(
      currentDragElementRef.current as HTMLDivElement
    );

    const dragContainerCoords = getCoords(
      dragContainerRef.current as HTMLDivElement
    );

    if (stickerCoords.left < dragContainerCoords.left) {
      currentDragElementRef.current.style.left = `${0}px`;
    }

    if (stickerCoords.top < dragContainerCoords.top) {
      currentDragElementRef.current.style.top = `${0}px`;
    }

    if (stickerCoords.left + stickerCoords.width > dragContainerCoords.right) {
      currentDragElementRef.current.style.left = `${
        dragContainerCoords.width - stickerCoords.width
      }px`;
    }

    if (stickerCoords.top + stickerCoords.width > dragContainerCoords.bottom) {
      currentDragElementRef.current.style.top = `${
        dragContainerCoords.height - stickerCoords.height
      }px`;
    }
  };

  /**
   * Use draggable areas hook
   */
  const { onMouseDown, onMouseMove } = useDraggableAreas({
    currentDragElementRef,
    dragContainerRef,
    onMountHandler: initDragAreas,
    onMouseMoveHandler: () => {
      detectParentContainer();

      correctPositionRelativeToParent();

      correctPositionRelativeToDragContainer();
    },
    onMouseDownHandler: () => {},
  });

  /**
   * Use sticker hook
   */
  const {
    closeModalHandler,
    createSticker,
    editSticker,
    editedSticker,
    isModalOpen,
    setModalVisibility,
    stickers,
    isDeleteMode,
    toggleDeleteMode,
    removeSticker,
  } = useSticker({
    board: BOARD_TYPE.left,
  });

  return (
    <>
      <Container
        className="sticker-container"
        onMouseMove={onMouseMove as VoidFunction}
        ref={dragContainerRef}
      >
        <div className="drag-area" ref={dragArea1Ref} />
        <div className="drag-area" ref={dragArea2Ref} />
        <div className="drag-area" ref={dragArea3Ref} />
        <Cell>
          <FloatingWrapper
            className="floating-wrapper"
            $paddingTop={80}
            ref={leftBannerWrapperRef}
          >
            <Banner $marginLeft={10}>
              <Image
                src="/banner1.jpeg"
                width={185}
                height={132}
                alt="Banner 1"
                priority
                draggable={false}
              />
            </Banner>
            <Banner $marginTop={10} $marginLeft={10}>
              <Image
                src="/banner2.jpeg"
                width={185}
                height={132}
                alt="Banner 2"
                draggable={false}
                priority
              />
            </Banner>
            <Banner $marginTop={10} $marginLeft={10}>
              <Image
                src="/banner3.jpeg"
                width={185}
                height={132}
                alt="Banner 3"
                draggable={false}
                priority
              />
            </Banner>
          </FloatingWrapper>
        </Cell>
        <Cell $justifyContent="center" $alignItems="start">
          <FloatingWrapper className="floating-wrapper" ref={logoWrapperRef}>
            <LogoWrapper>
              <Image
                src="/gerb-1.png"
                width={39}
                height={52}
                alt="Picture of the author"
                draggable={false}
              />
              <span>sÈcuritÈ</span>
            </LogoWrapper>
          </FloatingWrapper>
        </Cell>
        <Cell $alignItems="start" $justifyContent="flex-end">
          <FloatingWrapper className="floating-wrapper" ref={buttonsWrapperRef}>
            <ButtonsAddDelete
              add={setModalVisibility(true)}
              remove={toggleDeleteMode}
              marginTop={22}
              marginRight={16}
            />
          </FloatingWrapper>
        </Cell>
        <Cell />
        <Cell />
        <Cell />
        <Cell />
        <Cell />
        <Cell $alignItems="end" $justifyContent="flex-end">
          <FloatingWrapper
            className="floating-wrapper"
            ref={lastBannerWrapperRef}
          >
            <Banner
              $marginRight={10}
              $marginBottom={10}
              $width={129}
              $height={188}
            >
              <Image
                src="/banner4.jpeg"
                width={129}
                height={188}
                alt="Banner 4"
                draggable={false}
              />
            </Banner>
          </FloatingWrapper>
        </Cell>
        {stickers.map((item) => (
          <StickerComponent
            onMouseDown={onMouseDown as VoidFunction}
            key={item.id}
            sticker={item}
            editSticker={editSticker(item)}
            isDeleteMode={isDeleteMode}
            removeSticker={removeSticker(item.id)}
          />
        ))}
      </Container>
      {isModalOpen && (
        <ModalAddSticker
          closeModal={closeModalHandler}
          createSticker={createSticker}
          board={BOARD_TYPE.left}
          sticker={editedSticker}
        />
      )}
    </>
  );
}
