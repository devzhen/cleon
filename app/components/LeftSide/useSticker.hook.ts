import { useState } from "react";
import { v4 as uuid } from "uuid";

import { stickers as stickersArray } from "@/app//data";
import { DEFAULT_COORDS } from "@/app/constants";

const useSticker = () => {
  const [stickers, setStickers] = useState(stickersArray);

  /**
   * Add sticker
   */
  const addSticker = () => {
    const sticker = {
      id: uuid(),
      top: DEFAULT_COORDS.top,
      left: DEFAULT_COORDS.left,
      createdAt: new Date().toISOString(),
      text: "",
    };

    setStickers((prev) => [...prev, sticker]);
  };

  return {
    stickers,
    addSticker,
  }
}

export default useSticker