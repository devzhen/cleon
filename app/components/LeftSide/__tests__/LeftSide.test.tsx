import { render } from '@testing-library/react';

import { firstId, stickers } from '@/app/data';

import LeftSide from '..';

describe('LeftSide tests', () => {
  const props = {
    initialStickers: {
      [firstId]: stickers[firstId],
    },
  };
  it('matches the snapshot', () => {
    const { container } = render(<LeftSide {...props} />);

    expect(container).toMatchSnapshot();
  });
});
