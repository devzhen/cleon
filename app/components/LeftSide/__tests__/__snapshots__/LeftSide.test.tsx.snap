// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LeftSide tests matches the snapshot 1`] = `
<div>
  <div
    class="leftSide sticker-container"
    data-board="left"
  >
    <div
      data-role="drag-area"
      style="top: 0px; left: 0px; width: -6px; height: -6px;"
    />
    <div
      data-role="drag-area"
      style="top: 0px; left: 0px; width: -6px; height: -6px;"
    />
    <div
      data-role="drag-area"
      style="top: 0px; left: 0px; width: -6px; height: -6px;"
    />
    <div
      class="floatingWrapper"
    >
      <div
        class="leftBannerWrapper"
      >
        <div
          class="banner"
        >
          <img
            alt="Banner 1"
            data-nimg="1"
            decoding="async"
            draggable="false"
            fetchpriority="high"
            height="132"
            src="/_next/image?url=%2Fbanner1.jpeg&w=384&q=75"
            srcset="/_next/image?url=%2Fbanner1.jpeg&w=256&q=75 1x, /_next/image?url=%2Fbanner1.jpeg&w=384&q=75 2x"
            style="color: transparent;"
            width="185"
          />
        </div>
        <div
          class="banner"
        >
          <img
            alt="Banner 2"
            data-nimg="1"
            decoding="async"
            draggable="false"
            fetchpriority="high"
            height="132"
            src="/_next/image?url=%2Fbanner2.jpeg&w=384&q=75"
            srcset="/_next/image?url=%2Fbanner2.jpeg&w=256&q=75 1x, /_next/image?url=%2Fbanner2.jpeg&w=384&q=75 2x"
            style="color: transparent;"
            width="185"
          />
        </div>
        <div
          class="banner"
        >
          <img
            alt="Banner 3"
            data-nimg="1"
            decoding="async"
            draggable="false"
            fetchpriority="high"
            height="132"
            src="/_next/image?url=%2Fbanner3.jpeg&w=384&q=75"
            srcset="/_next/image?url=%2Fbanner3.jpeg&w=256&q=75 1x, /_next/image?url=%2Fbanner3.jpeg&w=384&q=75 2x"
            style="color: transparent;"
            width="185"
          />
        </div>
      </div>
    </div>
    <div
      class="logoWrapper"
    >
      <div
        class="floatingWrapper"
      >
        <div>
          <img
            alt="Picture of the author"
            data-nimg="1"
            decoding="async"
            draggable="false"
            height="52"
            loading="lazy"
            src="/_next/image?url=%2Fgerb-1.png&w=96&q=75"
            srcset="/_next/image?url=%2Fgerb-1.png&w=48&q=75 1x, /_next/image?url=%2Fgerb-1.png&w=96&q=75 2x"
            style="color: transparent;"
            width="39"
          />
          <span>
            sÈcuritÈ
          </span>
        </div>
      </div>
    </div>
    <div
      class="buttonsWrapper"
    >
      <div
        class="floatingWrapper"
      >
        <div
          class="container"
          style="margin-top: 22px; margin-right: 16px;"
        >
          <button>
            +
          </button>
          <button>
            -
          </button>
        </div>
      </div>
    </div>
    <div />
    <div />
    <div />
    <div />
    <div />
    <div
      class="bottomBannerWrapper"
    >
      <div
        class="floatingWrapper"
      >
        <div
          class="banner bottomBanner"
        >
          <img
            alt="Banner 4"
            data-nimg="1"
            decoding="async"
            draggable="false"
            height="188"
            loading="lazy"
            src="/_next/image?url=%2Fbanner4.jpeg&w=384&q=75"
            srcset="/_next/image?url=%2Fbanner4.jpeg&w=256&q=75 1x, /_next/image?url=%2Fbanner4.jpeg&w=384&q=75 2x"
            style="color: transparent;"
            width="129"
          />
        </div>
      </div>
    </div>
    <div
      class="container sticker"
      data-board="left"
      data-id="uuid/v4"
      role="presentation"
      style="top: 449px; left: 963px; z-index: 3;"
    >
      <span
        data-role="date"
      >
        02-03-2024 09:10:11 AM
      </span>
      <div
        data-role="content-wrapper"
      >
        <span>
          Draggable element 4
        </span>
      </div>
      <div
        data-role="image-wrapper"
      >
        <img
          alt="Picture of the author"
          data-nimg="1"
          decoding="async"
          draggable="false"
          height="13"
          loading="lazy"
          src="/_next/image?url=%2Freg.png&w=32&q=75"
          srcset="/_next/image?url=%2Freg.png&w=16&q=75 1x, /_next/image?url=%2Freg.png&w=32&q=75 2x"
          style="color: transparent;"
          width="13"
        />
      </div>
    </div>
  </div>
</div>
`;
